crates/shared/src/price_estimation/oneinch.rs

```diff
@@ -0,0 +1,318 @@
+use super::gas;
+use crate::oneinch_api::{OneInchClient, RestResponse, SellOrderQuoteQuery};
+use crate::price_estimation::{Estimate, PriceEstimating, PriceEstimationError, Query};
+use anyhow::Result;
+use cached::{Cached, TimedSizedCache};
+use futures::future;
+use model::order::OrderKind;
+use primitive_types::U256;
+use std::sync::{Arc, Mutex};
+
+pub struct OneInchPriceEstimator {
+    api: Arc<dyn OneInchClient>,
+    disabled_protocols: Vec<String>,
+    allowed_protocols: Arc<Mutex<TimedSizedCache<Vec<String>, Vec<String>>>>,
+}
+
+impl OneInchPriceEstimator {
+    async fn estimate(&self, query: &Query) -> Result<Estimate, PriceEstimationError> {
+        if query.kind == OrderKind::Buy {
+            return Err(PriceEstimationError::UnsupportedOrderType);
+        }
+
+        let quote = self
+            .api
+            .get_sell_order_quote(SellOrderQuoteQuery {
+                from_token_address: query.sell_token,
+                to_token_address: query.buy_token,
+                amount: query.in_amount,
+                protocols: self.get_protocol_argument().await?,
+                fee: None,
+                gas_limit: None,
+                connector_tokens: None,
+                complexity_level: None,
+                main_route_parts: None,
+                virtual_parts: None,
+                parts: None,
+                gas_price: None,
+            })
+            .await
+            .map_err(PriceEstimationError::Other)?;
+
+        match quote {
+            RestResponse::Ok(quote) => Ok(Estimate {
+                out_amount: quote.to_token_amount,
+                gas: U256::from(gas::SETTLEMENT_OVERHEAD) + quote.estimated_gas,
+            }),
+            RestResponse::Err(e) => {
+                Err(PriceEstimationError::Other(anyhow::anyhow!(e.description)))
+            }
+        }
+    }
+
+    pub fn new(api: Arc<dyn OneInchClient>, disabled_protocols: Vec<String>) -> Self {
+        Self {
+            api,
+            disabled_protocols,
+            allowed_protocols: Arc::new(Mutex::new(
+                TimedSizedCache::with_size_and_lifespan_and_refresh(1, 60, false),
+            )),
+        }
+    }
+
+    async fn get_protocol_argument(&self) -> Result<Option<Vec<String>>> {
+        if self.disabled_protocols.is_empty() {
+            return Ok(None);
+        }
+
+        if let Some(allowed_protocols) = self
+            .allowed_protocols
+            .lock()
+            .unwrap()
+            .cache_get(&self.disabled_protocols)
+        {
+            return Ok(Some(allowed_protocols.clone()));
+        }
+
+        let current_protocols = self.api.get_protocols().await?.protocols;
+        let allowed_protocols: Vec<String> = current_protocols
+            .into_iter()
+            // linear search through the Vec is okay because it's very small
```

Comment:
I'm willing to bet this is a tiny fraction of the total time taken to complete this method (most of it being waiting on I/O).

Reply:
Sure, but I still like to be aware of the runtime of the CPU bound code.

Reply:
I was mostly agreeing with your statement 😛. That the runtime of the code was fine in this instance.